tensorflow/compiler/mlir/hlo/lib/Dialect/mhlo/transforms/legalize_to_linalg.cc

```diff
@@ -224,6 +224,296 @@ static bool HasCanonicalDimensionNumbers(
+      if (lhs_shape[dim_ind_pos] != ShapedType::kDynamicSize) continue;
+      dim_size = b.create<memref::DimOp>(loc, lhs, dim_ind_pos);
+    } else {
+      // query from rhs vars
+      dim_ind_it = std::find(rhs_loop_vec.begin(), rhs_loop_vec.end(), dim_ind);
+      auto dim_ind_pos = dim_ind_it - rhs_loop_vec.begin();
+      auto rhs_shape = rhs.getType().dyn_cast<RankedTensorType>().getShape();
+      if (rhs_shape[dim_ind_pos] != ShapedType::kDynamicSize) continue;
+      dim_size = b.create<memref::DimOp>(loc, rhs, dim_ind_pos);
+    }
+    dyn_sizes.push_back(dim_size);
+  }
+  return dyn_sizes;
+}
+
+// Adds indices/axes that are missing from output set
+llvm::SmallSetVector<StringRef, 4> FindSummationAxes(
+    const llvm::SmallSetVector<StringRef, 4>& input_set,
+    const llvm::SmallSetVector<StringRef, 4>& output_set) {
+  llvm::SmallSetVector<StringRef, 4> summation_axes;
+  for (const StringRef& ind : input_set) {
+    if (!output_set.contains(ind)) summation_axes.insert(ind);
+  }
+  return summation_axes;
+}
+
+// Given a 1:1 map from std::string -> affine dimension expression
+// we can get the affine expression of dimensions that an
+// operand will access based on the input_str of einsum_config
+// For example:
+// let string_dim_umap = {'a' : d0, 'b' : d1, 'c' : d2}
+// for einsum_config "abc,cb->acb"
+// the first_input_operand will get umap[{"a","b","c"}] -> (d0, d1, d2)
+// the second_input_operand will get umap[{"c","b"}] -> (d2, d1)
+// the ouput_operand will get umap[{"a","c","b"}] -> (d0, d2, d1)
+SmallVector<AffineExpr> GetExprFromConfig(
+    const SmallVector<std::string>& loop_dims,
+    const DenseMap<StringRef, AffineExpr>& str_affine_dim_umap) {
+  SmallVector<AffineExpr> exprs;
+  for (const StringRef& dim : loop_dims) {
+    exprs.push_back(str_affine_dim_umap.lookup(dim));
+  }
+  return exprs;
+}
+
+// Convert the representation from string/vector<char> to vector<string>
+// i.e ("abc") -> {"a", "b", "c"}. For cases with ellipsis with batch rank 3:
+// get loop_dim = f("ab...cde") = {"a","b","0","1","2","c","d","e"}
+SmallVector<std::string> GetEinsumConfigAsVector(StringRef loop,
+                                                 const size_t& operand_rank) {
+  const std::string kEllipsis = "...";
+  SmallVector<std::string> loop_dim;
+  size_t pre_elip = loop.find(kEllipsis);
+  bool has_elip = pre_elip != std::string::npos;
+  if (!has_elip) pre_elip = loop.size();
+  // Add the dimension until the end or up to ellipsis if it exist
+  for (int pre_elip_ind = 0; pre_elip_ind < pre_elip; pre_elip_ind++)
+    loop_dim.push_back(std::string(1, loop[pre_elip_ind]));
+  if (!has_elip) return loop_dim;
+  // Case where Ellipsis presence:
+  size_t non_batch_rank = loop.size() - kEllipsis.size();
+  size_t batch_rank = operand_rank - non_batch_rank;
+  // Add the batch dimension ("0",...,"N") where N is rank of batch into the
+  // loop
```

Comment:
Add a period at the end.

```suggestion
  // loop.
```

Reply:
done